lib/SILOptimizer/Transforms/TempRValueElimination.cpp

```diff
@@ -347,6 +347,49 @@ SILInstruction *TempRValueOptPass::getLastUseWhileSourceIsNotModified(
   return nullptr;
 }
 
+/// Tries to move an end_access down to extend the access scope over all uses
+/// of the temporary. For example:
+///
+///   %a = begin_access %src
+///   copy_addr %a to [initialization] %temp : $*T
+///   end_access %a
+///   use %temp
+///
+/// We must not replace %temp with %a after the end_access. Instead we try to
+/// move the end_access after "use %temp".
+bool TempRValueOptPass::extendAccessScopes(
+    CopyAddrInst *copyInst, SILInstruction *lastUseInst) {
+
+  SILValue copySrc = copyInst->getSrc();
+  EndAccessInst *endAccessToMove = nullptr;
+  auto begin = std::next(copyInst->getIterator());
+  auto end = std::next(lastUseInst->getIterator());
+
+  for (SILInstruction &inst : make_range(begin, end)) {
+    if (auto *endAccess = dyn_cast<EndAccessInst>(&inst)) {
+      // Is this the end of an access scope of the copy-source?
+      if (!aa->isNoAlias(copySrc, endAccess->getSource())) {
+        // To keep things simple, we can just move a single end_access.
+        if (endAccessToMove)
+          return false;
+        endAccessToMove = endAccess;
+      }
+    } else if (endAccessToMove) {
+      // We cannot move an end_access over a begin_access. This would destroy
+      // the proper nesting of accesses.
+      if (isa<BeginAccessInst>(&inst))
+        return false;
+      // Don't extend a read-access scope over a (potential) write.
+      if (aa->mayWriteToMemory(&inst, endAccessToMove->getSource()))
```

Comment:
@eeckstein looking at this again I'm not sure about it

- minor point: we don't actually check that this is a "read" access. We can merge read/write accesses, so this copy_addr may be within a write

- either way, the idea with access markers is that we don't care about other memory accesses themselves, only other access scopes. So we shouldn't need to look for other "writes"

- we do need to look for other access scopes, either read or writes. The code checks for begin_access above, but an access could be in any call (unless it has no reads or writes).

Reply:
* MemBehavior yields a  ``MayWrite`` for the ``begin_access`` and ``end_access`` instructions of a modify (or init) access. Therefore it can only be a read ``end_access`` what we see here. For a modify ``end_access`` we would have bailed earlier. I'll add a comment and a test.

* I need to use alias analysis to do the check (instead of just looking for access markers), because of function calls. As you said, ``inst`` could be a function call containing an access scope. 

* It's fine if ``inst`` is a function call with an access scope. Moving the ``end_access`` above a call will not destroy the proper nesting. And it's guaranteed that in this case both accesses are read-only accesses.

Reply:
I looked at the latest code and see asserts and comments that cover all my points. Thanks.